### PR TITLE
onAction

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Changes

Still need to write tests, just putting this out because @samwinslow wanted to see it.

This came out of a pairing session with @mariusandra 

## How did you test this code?

I'm still writing tests. Tested it manually though and it works nicely :D

### CODE REVIEW

File: plugin-server/src/types.ts [hunk 1]
```diff
@@ -328,6 +329,7 @@ export type VMMethods = {
     setupPlugin?: () => Promise<void>
     teardownPlugin?: () => Promise<void>
     onEvent?: (event: PluginEvent) => Promise<void>
+    onAction?: (action: Action, event: PluginEvent) => Promise<void>
```
Comment (yakkomajuri) [inline plugin-server/src/types.ts:332]: essentially this is what you should care about @samwinslow

File: plugin-server/src/main/ingestion-queues/ingest-event.ts [hunk 2]
```diff
@@ -48,6 +53,20 @@ export async function ingestEvent(
...
+                    runInstrumentedFunction({
+                        server,
+                        event: processedEvent!,
+                        func: (event) => workerMethods.onAction(actionMatch, event),
+                        statsKey: `kafka_queue.on_action`,
+                        timeoutMessage: 'After 30 seconds still running onAction',
+                    })
+                )
+            )
+        }
```
Comment (mariusandra) [inline plugin-server/src/main/ingestion-queues/ingest-event.ts:69]: I wonder if changing the flow from:

```
sequential(
  parallel(IngestEvent, onEvent),
  onAction,
)
```

To:

```
parallel(
  sequential(IngestEvent, onAction), 
  onEvent
)
```

makes any sense? 

Probably not, as I expect `ingestEvent` to take much much longer 99.99% of the time than `onEvent`, which normally uses a buffer.

**Review by mariusandra — COMMENTED** [review]
This is an amazing turnaround time!

Now the question is: should each plugin get every action, or do we control the flow somehow? Ideally we'd like to save as many CPU cycles as possible.

**Comment by yakkomajuri** [comment]
Finally had a bit of time to write this test.

cc @mariusandra and also maybe @samwinslow you want to try to review this too?

**Comment by yakkomajuri** [comment]
@mariusandra the ideal solution would be to pick plugins to run on the action page. I definitely don't have time to build this anytime soon but maybe @samwinslow would like to give this a shot? I'm happy to support it.

File: plugin-server/src/worker/plugins/run.ts [hunk 3]
```diff
@@ -31,6 +32,27 @@ export async function runOnEvent(server: Hub, event: PluginEvent): Promise<void>
...
+            if (onAction) {
+                const timer = new Date()
+                try {
+                    await onAction(action, event)
+                } catch (error) {
+                    await processError(server, pluginConfig, error, event)
+                    server.statsd?.increment(`plugin.${pluginConfig.plugin?.name}.on_action.ERROR`)
+                }
+                server.statsd?.timing(`plugin.${pluginConfig.plugin?.name}.on_action`, timer)
+                //captureTimeSpentRunning(event.team_id, timer, 'onAction')
```
Comment (mariusandra) [inline plugin-server/src/worker/plugins/run.ts:50]: `//`?

File: plugin-server/src/main/ingestion-queues/ingest-event.ts [hunk 4]
```diff
@@ -48,6 +53,20 @@ export async function ingestEvent(
                 timeoutMessage: `After 30 seconds still running ${isSnapshot ? 'onSnapshot' : 'onEvent'}`,
             }),
         ])
+
+        if (actionMatches.length > 0) {
+            await Promise.all(
+                actionMatches.map((actionMatch) =>
+                    runInstrumentedFunction({
+                        server,
+                        event: processedEvent!,
```
Comment (mariusandra) [inline plugin-server/src/main/ingestion-queues/ingest-event.ts:62]: Should we get rid of the `!` somehow?

**Review by mariusandra — APPROVED** [review]
LGTM for a first implementation. We should put some work into mapping plugins with actions like you said, but for now a simple "if" statement inside the plugin will do the trick if someone needs to limit its scope...

File: plugin-server/src/worker/plugins/run.ts (hunk 3 above)
Comment (yakkomajuri) [reply plugin-server/src/worker/plugins/run.ts:50]: lingered

File: plugin-server/src/main/ingestion-queues/ingest-event.ts (hunk 4 above)
Comment (yakkomajuri) [reply plugin-server/src/main/ingestion-queues/ingest-event.ts:62]: lol I can try :D

File: plugin-server/src/worker/plugins/run.ts (hunk 3 above)
Comment (yakkomajuri) [inline plugin-server/src/worker/plugins/run.ts:50]:
```suggestion
                captureTimeSpentRunning(event.team_id, timer, 'onAction')
```

File: plugin-server/src/main/ingestion-queues/ingest-event.ts (hunk 4 above)
Comment (yakkomajuri) [reply plugin-server/src/main/ingestion-queues/ingest-event.ts]: done, this was weird.

Comment (mariusandra) [reply plugin-server/src/main/ingestion-queues/ingest-event.ts]: lol :D 

I guess you can never know if a mutable value is still there when a callback runs...

**Pull request merged**